File: day11_part1_2016.py

import os
import sys
import string

def rtgHellDay(input_str, part):
    def newInitialState(input):
        s = {
            "floors": [[] for _ in range(4)],
            "elevatorLevel": 0,
            "steps": 0
        }

        for lineIndex, line in enumerate(input.split("\n")):
            parts = [v.strip(",.") for v in line.split(" ")]
            for i, word in enumerate(parts):
                if word == "generator":
                    material = parts[i-1]
                    s["floors"][lineIndex].append({"isChip": False, "material": material})
                elif word == "microchip":
                    material = parts[i-1][:parts[i-1].index("-comp")]
                    s["floors"][lineIndex].append({"isChip": True, "material": material})

        return s

    def hashKey(s):
        mapGenToIndex = {}
        mapChipToIndex = {}
        for flIndex, fl in enumerate(s["floors"]):
            for half in fl:
                if half["isChip"]:
                    mapChipToIndex[half["material"]] = flIndex
                else:
                    mapGenToIndex[half["material"]] = flIndex

        genChipPairs = []
        for material in mapGenToIndex.keys():
            genChipPairs.append([mapGenToIndex[material], mapChipToIndex[material]])

        genChipPairs.sort(key=lambda x: (x[0], x[1]))

        return str(s["elevatorLevel"]) + str(genChipPairs)

    def isValid(s):
        for i, fl in enumerate(s["floors"]):
            gensSeen = {}
            for half in fl:
                if not half["isChip"]:
                    gensSeen[half["material"]] = True

            if len(gensSeen) == 0:
                continue

            for half in fl:
                if half["isChip"] and not gensSeen.get(half["material"], False):
                    return False

        return True

    def isDone(s):
        lenSum = sum([len(fl) for fl in s["floors"][:3]])
        return lenSum == 0

    def getMovablePermIndices(s):
        permsToMove = []

        currentLevel = s["floors"][s["elevatorLevel"]]

        for i in range(len(currentLevel)):
            for j in range(i+1, len(currentLevel)):
                permsToMove.append([i, j])

        for i in range(len(currentLevel)):
            permsToMove.append([i])

        return permsToMove

    def clone(s):
        cl = {
            "floors": [fl[:] for fl in s["floors"]],
            "elevatorLevel": s["elevatorLevel"],
            "steps": s["steps"]
        }
        return cl

    def getNextStates(s):
        futureStates = []
        movablePermIndices = getMovablePermIndices(s)
        eleDiffs = [1] if s["elevatorLevel"] < 3 else []
        eleDiffs += [-1] if s["elevatorLevel"] > 0 else []

        for eleDiff in eleDiffs:
            for permIndices in movablePermIndices:
                cl = clone(s)
                cl["elevatorLevel"] += eleDiff
                cl["steps"] += 1
                oldLevel = s["elevatorLevel"]
                newLevel = cl["elevatorLevel"]

                for index in permIndices:
                    cl["floors"][newLevel].append(s["floors"][oldLevel][index])

                for in_ in range(len(permIndices)-1, -1, -1):
                    cl["floors"][oldLevel][permIndices[in_]] = cl["floors"][oldLevel][-1]
                    cl["floors"][oldLevel] = cl["floors"][oldLevel][:-1]

                if isValid(cl):
                    futureStates.append(cl)

        return futureStates

    currentState = newInitialState(input_str)

    if part == 2:
        currentState["floors"][0] += [
            {"isChip": False, "material": "elerium"},
            {"isChip": True, "material": "elerium"},
            {"isChip": False, "material": "dilithium"},
            {"isChip": True, "material": "dilithium"}
        ]

    queue = [currentState]
    prevStates = {}
    while len(queue) > 0:
        front = queue[0]
        queue = queue[1:]

        if isDone(front):
            return front["steps"]

        hash_key = hashKey(front)
        if hash_key in prevStates:
            continue
        prevStates[hash_key] = True

        nextStates = getNextStates(front)
        queue += nextStates

    return -1

with open("input.txt", "r") as file:
    input_data = file.read()

part = 1
ans = rtgHellDay(input_data, part)
print(ans)

File: objectivec/day11_part1_2016.m

#import <Foundation/Foundation.h>

@interface Half : NSObject
@property (nonatomic, assign) BOOL isChip;
@property (nonatomic, strong) NSString *material;
- (instancetype)initWithIsChip:(BOOL)isChip material:(NSString *)material;
- (NSString *)description;
@end

@implementation Half
- (instancetype)initWithIsChip:(BOOL)isChip material:(NSString *)material {
    self = [super init];
    if (self) {
        _isChip = isChip;
        _material = material;
    }
    return self;
}
- (NSString *)description {
    return [NSString stringWithFormat:@"%@ %@", _material, _isChip ? @"microchip" : @"generator"];
}
@end

@interface State : NSObject
@property (nonatomic, strong) NSMutableArray<NSMutableArray<Half *> *> *floors;
@property (nonatomic, assign) NSInteger elevatorLevel;
@property (nonatomic, assign) NSInteger steps;
- (instancetype)initWithInput:(NSString *)input;
- (NSString *)description;
- (NSString *)hashKey;
- (BOOL)isValid;
- (BOOL)isDone;
- (NSArray<State *> *)getNextStates;
- (State *)clone;
- (NSArray<NSArray<NSNumber *> *> *)getMovablePermIndices;
@end

@implementation State
- (instancetype)initWithInput:(NSString *)input {
    self = [super init];
    if (self) {
        _floors = [NSMutableArray arrayWithCapacity:4];
        for (int i = 0; i < 4; i++) {
            [_floors addObject:[NSMutableArray array]];
        }
        NSArray *lines = [input componentsSeparatedByString:@"\n"];
        for (NSInteger lineIndex = 0; lineIndex < lines.count; lineIndex++) {
            NSString *line = lines[lineIndex];
            NSArray *parts = [line componentsSeparatedByString:@" "];
            NSMutableArray *trimmedParts = [NSMutableArray arrayWithCapacity:parts.count];
            for (NSString *part in parts) {
                [trimmedParts addObject:[part stringByTrimmingCharactersInSet:[NSCharacterSet characterSetWithCharactersInString:@",."]]];
            }
            for (NSInteger i = 0; i < trimmedParts.count; i++) {
                NSString *word = trimmedParts[i];
                if ([word isEqualToString:@"generator"]) {
                    NSString *material = trimmedParts[i - 1];
                    [_floors[lineIndex] addObject:[[Half alloc] initWithIsChip:NO material:material]];
                } else if ([word isEqualToString:@"microchip"]) {
                    NSString *material = [trimmedParts[i - 1] substringToIndex:[trimmedParts[i - 1] rangeOfString:@"-comp"].location];
                    [_floors[lineIndex] addObject:[[Half alloc] initWithIsChip:YES material:material]];
                }
            }
        }
        _elevatorLevel = 0;
        _steps = 0;
    }
    return self;
}

- (NSString *)description {
    NSMutableString *sb = [NSMutableString string];
    [sb appendFormat:@"Level %ld x Steps %ld\n", (long)_elevatorLevel, (long)_steps];
    for (NSInteger i = 0; i < _floors.count; i++) {
        [sb appendFormat:@"  %ld: %@\n", (long)i, _floors[i]];
    }
    return sb;
}

- (NSString *)hashKey {
    NSMutableDictionary<NSString *, NSNumber *> *mapGenToIndex = [NSMutableDictionary dictionary];
    NSMutableDictionary<NSString *, NSNumber *> *mapChipToIndex = [NSMutableDictionary dictionary];
    for (NSInteger flIndex = 0; flIndex < _floors.count; flIndex++) {
        NSArray<Half *> *fl = _floors[flIndex];
        for (Half *half in fl) {
            if (half.isChip) {
                mapChipToIndex[half.material] = @(flIndex);
            } else {
                mapGenToIndex[half.material] = @(flIndex);
            }
        }
    }
    NSMutableArray<NSArray<NSNumber *> *> *genChipPairs = [NSMutableArray array];
    for (NSString *material in mapGenToIndex) {
        [genChipPairs addObject:@[mapGenToIndex[material], mapChipToIndex[material]]];
    }
    [genChipPairs sortUsingComparator:^NSComparisonResult(NSArray<NSNumber *> *obj1, NSArray<NSNumber *> *obj2) {
        if ([obj1[0] integerValue] != [obj2[0] integerValue]) {
            return [obj1[0] compare:obj2[0]];
        }
        return [obj1[1] compare:obj2[1]];
    }];
    return [NSString stringWithFormat:@"%ld%@", (long)_elevatorLevel, genChipPairs];
}

- (BOOL)isValid {
    for (NSArray<Half *> *floor in _floors) {
        NSMutableSet<NSString *> *gensSeen = [NSMutableSet set];
        for (Half *half in floor) {
            if (!half.isChip) {
                [gensSeen addObject:half.material];
            }
        }
        if (gensSeen.count == 0) {
            continue;
        }
        for (Half *half in floor) {
            if (half.isChip && ![gensSeen containsObject:half.material]) {
                return NO;
            }
        }
    }
    return YES;
}

- (BOOL)isDone {
    NSInteger lenSum = 0;
    for (NSInteger i = 0; i < 3; i++) {
        lenSum += _floors[i].count;
    }
    return lenSum == 0;
}

- (NSArray<NSArray<NSNumber *> *> *)getMovablePermIndices {
    NSMutableArray<NSMutableArray<NSNumber *> *> *permsToMove = [NSMutableArray array];
    NSArray<Half *> *currentLevel = _floors[_elevatorLevel];
    for (NSInteger i = 0; i < currentLevel.count; i++) {
        for (NSInteger j = i + 1; j < currentLevel.count; j++) {
            [permsToMove addObject:[NSMutableArray arrayWithObjects:@(i), @(j), nil]];
        }
    }
    for (NSInteger i = 0; i < currentLevel.count; i++) {
        [permsToMove addObject:[NSMutableArray arrayWithObject:@(i)]];
    }
    return permsToMove;
}

- (State *)clone {
    State *cl = [[State alloc] init];
    cl.elevatorLevel = _elevatorLevel;
    cl.steps = _steps;
    cl.floors = [NSMutableArray arrayWithCapacity:4];
    for (NSArray<Half *> *floor in _floors) {
        [cl.floors addObject:[NSMutableArray arrayWithArray:floor]];
    }
    return cl;
}

- (NSArray<State *> *)getNextStates {
    NSMutableArray<State *> *futureStates = [NSMutableArray array];
    NSArray<NSArray<NSNumber *> *> *movablePermIndices = [self getMovablePermIndices];
    NSArray<NSNumber *> *eleDiffs;
    if (_elevatorLevel < _floors.count - 1) {
        eleDiffs = @[@1];
    }
    if (_elevatorLevel > 0) {
        if (eleDiffs) {
            eleDiffs = [eleDiffs arrayByAddingObject:@-1];
        } else {
            eleDiffs = @[@-1];
        }
    }
    if (!eleDiffs) {
        return futureStates;
    }
    for (NSNumber *eleDiff in eleDiffs) {
        for (NSArray<NSNumber *> *permIndices in movablePermIndices) {
            State *cl = [self clone];
            cl.elevatorLevel += [eleDiff integerValue];
            cl.steps++;
            NSInteger oldLevel = _elevatorLevel;
            NSInteger newLevel = cl.elevatorLevel;
            for (NSNumber *index in permIndices) {
                [cl.floors[newLevel] addObject:cl.floors[oldLevel][[index integerValue]]];
            }
            for (NSInteger in = permIndices.count - 1; in >= 0; in--) {
                NSInteger index = [permIndices[in] integerValue];
                cl.floors[oldLevel][index] = cl.floors[oldLevel].lastObject;
                [cl.floors[oldLevel] removeLastObject];
            }
            if ([cl isValid]) {
                [futureStates addObject:cl];
            }
        }
    }
    return futureStates;
}
@end

NSInteger rtgHellDay(NSString *input) {
    State *currentState = [[State alloc] initWithInput:input];
    NSMutableArray<State *> *queue = [NSMutableArray arrayWithObject:currentState];
    NSMutableDictionary<NSString *, NSNumber *> *prevStates = [NSMutableDictionary dictionary];
    while (queue.count > 0) {
        State *front = queue.firstObject;
        [queue removeObjectAtIndex:0];
        if ([front isDone]) {
            return front.steps;
        }
        NSString *hash = [front hashKey];
        if (prevStates[hash]) {
            continue;
        }
        prevStates[hash] = @(YES);
        NSArray<State *> *nextStates = [front getNextStates];
        [queue addObjectsFromArray:nextStates];
    }
    return -1;
}

NSString *readFile(NSString *path) {
    NSError *error;
    NSString *content = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:&error];
    if (error) {
        NSLog(@"Error reading file: %@", error);
        return nil;
    }
    return [content stringByTrimmingCharactersInSet:[NSCharacterSet newlineCharacterSet]];
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *filePath = [[NSBundle mainBundle] pathForResource:@"input" ofType:@"txt"];
        NSString *input = readFile(filePath);
        if (input) {
            NSInteger ans = rtgHellDay(input);
            NSLog(@"%ld", (long)ans);
        }
    }
    return 0;
}
